### ISSUE TITLE
Integrate thumbnail previews during the Wikimedia Commons batch file upload process

### ISSUE THREAD

**Issue by trnstlntk** [OpenRefine/OpenRefine#4277]
See https://commons.wikimedia.org/wiki/Commons_talk:OpenRefine

Request received in a conversation in the Wikimedia Commons Telegram channel. During a batch upload process of media files, it is extremely helpful if one can (easily) see thumbnail previews of the media files that are being uploaded.

Some existing Wikimedia Commons (batch) upload tools support this indeed (the default UploadWizard, for instance), others don't (Pattypan only shows previews of files and their infoboxes during the checking phase of the upload process, after all data has been prepared already).

OpenRefine is essentially a data-centric tool, so this may be a stretch, but it's good to have this request on the radar, as it makes a lot of sense IMO.

**Comment by thadguidry** [comment]
The extension can and should be developed to support what users need irregardless of OpenRefine. That's the beauty of an extension. Extending in ways the original program had no need for.

**Comment by wetneb** [comment]
@trnstlntk, @lozanaross and I have discussed this during a training workshop with users from cultural institutions, and one attendee asked if the feature could be available for projects that are unrelated to Wikimedia Commons.

The issue could indeed be formulated like this: say I have a dataset where one column contains local paths to images, I would like to have OpenRefine display small thumbnails of those images in the corresponding cells.

Given that we currently do not have any extension point which would let extensions customize cell rendering (as far as I can tell), it seems that we have two reasons to implement this in the core directly. I think this could go in some sort of view options of a given column, potentially?

**Comment by trnstlntk** [comment]
@lozanaross has created various wireframes that include file previews. A selection from the latest (April 26) version of the wireframes:

While working on a project:
![image](https://user-images.githubusercontent.com/9881645/165995117-6cecbb6b-7763-4755-8067-358d01e2f7a5.png)

We're also considering a functionality to make the thumbnail bigger here, especially useful when the user will be working with many similar media files that they want to compare:
![image](https://user-images.githubusercontent.com/9881645/165995622-927ef380-171d-4bed-a541-34348863ba49.png)
![image](https://user-images.githubusercontent.com/9881645/165995692-61a0f865-8c63-412f-ab53-c1ab6641fe8a.png)


During upload (schema) preview:
![image](https://user-images.githubusercontent.com/9881645/165995403-325ba2d3-16f3-4029-8f07-472719bf56b7.png)

**Comment by trnstlntk** [comment]
Assigning this task to @lozanaross (design; work in progress) and @j-sal (development of this feature). I didn't see an explicit deadline attached during the roadmapping exercise we did last week, but we placed this task around July so I'm provisionally adding a July 31 deadline.

**Comment by trnstlntk** [comment]
> @trnstlntk, @lozanaross and I have discussed this during a training workshop with users from cultural institutions, and one attendee asked if the feature could be available for projects that are unrelated to Wikimedia Commons.
> 
> The issue could indeed be formulated like this: say I have a dataset where one column contains local paths to images, I would like to have OpenRefine display small thumbnails of those images in the corresponding cells.
> 
> Given that we currently do not have any extension point which would let extensions customize cell rendering (as far as I can tell), it seems that we have two reasons to implement this in the core directly. I think this could go in some sort of view options of a given column, potentially?

I'll be happy to edit this issue to reflect the broader scope. We could also split it in seperate tasks?

For Wikimedia Commons, we identified distinct use cases that will need (as I understood it) different development approaches:

- [ ] User wants to use OpenRefine to add structured data to existing files from Wikimedia Commons. They load a series of file paths from Wikimedia Commons and then display thumbnails of these in OpenRefine. This can build further upon the work by @eugeneegbe on the [Commons Reconciliation Service](https://commonsreconcile.toolforge.org). (This is the first use case we want to tackle; deadline July 31)
- [ ] User wants to use OpenRefine to upload new files to Wikimedia Commons, from harddrive. They select a set of files from their harddrive. We need to develop a way to display thumbnails for these in OpenRefine.
- [ ] User wants to use OpenRefine to upload new files to Wikimedia Commons, from the web. They select a set of URLs (or start a project from API, or similar). We need to develop a way to display thumbnails for these in OpenRefine. 

For the second and third scenario: we placed work on this task around September 2022 in our roadmap. Due to time constraints perhaps we won't get there during the current grant.

**Comment by wetneb** [comment]
Yes we could potentially have three separate issues I think.

**Comment by lozanaross** [comment]
I think we discussed this and indeed we need separate issues, I'll set the others up - I think we also agreed the deadline for this first case (Commons images) is July 01 because we want it for wikimania?

**Comment by lozanaross** [comment]
The way the issue is framed at the moment makes me think it belongs in the extension repo, not in the main repo btw. I think it needs to be abstracted and possibly created later (after this issue is closed) in the main OpenRefine repo.

**Comment by trnstlntk** [comment]
> I think we discussed this and indeed we need separate issues, I'll set the others up - I think we also agreed the deadline for this first case (Commons images) is July 01 because we want it for wikimania?

We didn't put a deadline on that specific post it note 😅  but we indeed placed it right before Wikimania, so I'd say yes.

**Comment by lozanaross** [comment]
> The way the issue is framed at the moment makes me think it belongs in the extension repo, not in the main repo btw. I think it needs to be abstracted and possibly created later (after this issue is closed) in the main OpenRefine repo.

Replying to myself - this is a dumb comment because I realize you cannot move issues across repos in Github (unless I am just not seeing how?? it is possible in Gitlab). I guess we should keep this here for now, but it's technically part of the SDC extension because we're first developing the thumbs for commons only, right? Then later the options to render local / remote files?

**Comment by wetneb** [comment]
> I realize you cannot move issues across repos in Github

Yes you can, it is called "Transfer issue" on the right-hand side, and as a member of the organization it should be available to you.
